# Not yet implemented
- Notebook for ideas and functions related to cleaning the quarterly statements data

In [ ]:
balance = pd.read_csv('./data/raw/balance.csv').convert_dtypes()

Remove duplicates on selected columns

In [ ]:
balance = pd.read_csv('./data/raw/balance.csv').convert_dtypes()
dup_features = [
    "cik", 
    "calendarYear", 
    "totalAssets",
    "totalLiabilities",
    "totalEquity",
    ]
balance_clean = balance.drop_duplicates(subset = dup_features, keep='first')
balance_clean = balance_clean.reset_index(drop=True)

Manualy fix some periods

In [ ]:
mask = balance_clean.period == 'Q'
balance_clean.loc[mask, 'period'] = np.nan
balance_clean['period'] = balance_clean.period.str.replace('Q', '')
balance_clean.period = balance_clean.period.astype(pd.Int64Dtype())

Rename columns

In [ ]:
col_rename = {
    'fillingDate':'filingDate',
    'date':'reportDate',
    'period':'periodAPI',
}
balance_clean = balance_clean.rename(columns=col_rename)

join Balance and SEC admission dataframes to get the form columns to the balance sheet

In [ ]:
# reset_indexes to keep track
balance_clean = balance_clean.reset_index(drop=True)
sec = sec.reset_index(drop=True)

1st. join.
- reportDate, symbol, cik

In [ ]:
joincols = ['reportDate', 'cik']
secols = ['form']
# first join
first_join = balance_clean.merge(
    sec.loc[:, joincols + secols], 
    how='left', 
    on = joincols,
    )
mask = first_join.form.isna()
to_join = first_join[mask]
result_1 = first_join[~mask]
result_1.shape, to_join.shape

((43590, 55), (3841, 55))

2nd join.
- filingDate, symbol, cik

In [ ]:
joincols = ['filingDate', 'cik']
secols = ['form']
to_join = to_join.drop(columns=secols)
# first join
second_join = to_join.merge(
    sec.loc[:, joincols + secols], 
    how='left', 
    on = joincols,
    )
mask = second_join.form.isna()
to_join = second_join[mask]
result_2 = second_join[~mask]
result_2.shape, to_join.shape

((333, 55), (3508, 55))

In [ ]:
balance_clean.shape[0] - sec.shape[0]

2205

missing around 2400 vals, OK

In [ ]:
balance_clean = pd.concat([result_1, second_join])
balance_clean.head(2)

reportDate symbol reportedCurrency    cik  filingDate         acceptedDate  \
0  2021-09-30    MMM              USD  66740  2021-10-26  2021-10-26 14:53:46   
1  2021-06-30    MMM              USD  66740  2021-07-27  2021-07-27 14:55:50   

   calendarYear  periodAPI  cashAndCashEquivalents  shortTermInvestments  ...  \
0          2021          3            4878000000.0           855000000.0  ...   
1          2021          2            4695000000.0           805000000.0  ...   

   totalLiabilitiesAndStockholdersEquity  minorityInterest    totalEquity  \
0                          48208000000.0        70000000.0  14530000000.0   
1                          48239000000.0        68000000.0  14516000000.0   

   totalLiabilitiesAndTotalEquity  totalInvestments      totalDebt  \
0                   48278000000.0               0.0  19007000000.0   
1                   48307000000.0               0.0  19119000000.0   

         netDebt                                               link  \
0  14129000000.0  https://www.sec.gov/Archives/edgar/data/66740/...   
1  14424000000.0  https://www.sec.gov/Archives/edgar/data/66740/...   

                                           finalLink  form  
0  https://www.sec.gov/Archives/edgar/data/66740/...  10-Q  
1  https://www.sec.gov/Archives/edgar/data/66740/...  10-Q  

[2 rows x 55 columns]

EXPLORE: missing values
- Most of them are because the sec submissions API is missing older filings.
    - Checking if there is an issue with my gathering function

DONE
- There are no issues with the function, the problem is some companies changed their symbols and cik in the sec, such as google before 2015 was named GOOGLE .inc and its cik was 1288776

Since we cannot trust the API symbol,year,period nomenclature so we use the K from from the SEC as guidance and from there we expand it in all directions

In [ ]:
# set the column to infer as nan
try:
    balance_clean.insert(0, 'infered_sec', np.nan)
except:
    pass
# sort
balance_clean = balance_clean.sort_values(['symbol','reportDate'])
# do 10 times
for i in range(10):
    # init balance_clean on 4 if the form is a 10K statement
    cond = balance_clean.form.isin(['10-K', '10-K405', '10-KT'])
    balance_clean['infered_sec'] = balance_clean['infered_sec'].mask(cond= cond, other=4)
    # expand values vertically and horizontally
    for n in range(1,5):
        # only overwrite if its a nan
        nullcond = balance_clean.infered_sec.isna()
        # move the 4 within group and make mask
        cond1 = balance_clean.groupby('symbol').infered_sec.shift(n).eq(4)
        # set to value n if shifted mask & nan
        balance_clean['infered_sec'].mask(cond=(cond1 & nullcond), other=n, inplace = True)
        # shift same for the other direction
        cond2 = balance_clean.groupby('symbol').infered_sec.shift(-n).eq(4)
        if n != 4:
            balance_clean['infered_sec'].mask(cond=(nullcond & cond2), other=(4-n), inplace = True)
        else:
            balance_clean['infered_sec'].mask(cond=(nullcond & cond2), other=4, inplace = True)

In [ ]:
balance_clean.infered_sec.isna().sum()

927

Estimar los periodos a partir de la media de periodo, quitar los simbolos en los que ha habido un T y comprobar si hay algun form K con periodo no 4

In [ ]:
dates_processing(periods,'reportDate', 'filingDate')
# get the dictionary containing the circular mean by period and symbol
dict_means_date = grouped_circmean(periods, 'reportDate_yday', low=0, high=365, levels = ['symbol', 'infered_sec'])
infered_period_date = periods.apply(lambda x: infer_period(x['symbol'], x['reportDate_yday'], dict_means_date), axis=1)
dict_means_fillingDate = grouped_circmean(periods, 'filingDate_yday', low=0, high=365, levels = ['symbol', 'infered_sec'])
infered_period_filling_date = periods.apply(lambda x: infer_period(x['symbol'], x['filingDate_yday'], dict_means_fillingDate), axis=1)

In [ ]:
periods['infered_reportDate'] = infered_period_date
periods['infered_filingDate'] = infered_period_filling_date

In [ ]:
cond1 = periods.infered_sec != periods.infered_reportDate
cond2 = periods.infered_sec != periods.infered_filingDate
supercond = cond1 & cond2
subset = periods[supercond]

In [ ]:
subset.symbol.unique()

array(['AAPL', 'ADM', 'AEP', 'APD', 'ATVI', 'BKR', 'BLL', 'BXP', 'CB',
       'CSCO', 'DAL', 'DG', 'GD', 'GPN', 'HCA', 'IT', 'JCI', 'KIM', 'KR',
       'LEN', 'LHX', 'MOS', 'MTB', 'ODFL', 'OKE', 'ROK', 'ROP', 'SEDG',
       'STZ', 'SWK', 'SWKS', 'TTWO', 'UAL', 'VFC', 'WAB'], dtype=object)

If a symbol had a 10-KT form or 10-QT form, it means it had altered its fiscal year, and the period inference method is not significant

In [ ]:
# get all the observations whose form is either 10-KT or 10-QT
mask_form = periods.form.isin(['10-KT', '10-QT'])
# get their symbols
symbols = periods[mask_form].symbol
# remove them from the subset
subset = subset[~subset.symbol.isin(symbols)]

For the remaining subset keep the estimated values

In [ ]:
periods.loc[subset.index,'infered_sec'] = periods.loc[subset.index,'infered_reportDate']

Now that the sec submissions df is mostly clean we join it with the balance sheet. Because all data sources are dirty, we're left joining the balance sheet and consecuently update the resulting DataFrame to minimize the amount of observations considered null because of dirty features.

In [ ]:
balance_clean['reportDate'] = pd.to_datetime(balance_clean['reportDate'])
balance_clean['filingDate'] = pd.to_datetime(balance_clean['filingDate'])

In [ ]:
balance_clean = dates_processing(balance_clean, 'reportDate', 'filingDate')

In [ ]:
periods.columns

Index(['filingDate', 'reportDate', 'symbol', 'cik', 'form', 'infered_sec',
       'reportDate_int', 'reportDate_month', 'reportDate_year',
       'reportDate_yday', 'filingDate_int', 'filingDate_month',
       'filingDate_year', 'filingDate_yday', 'infered_reportDate',
       'infered_filingDate'],
      dtype='object')

In [ ]:
balance_clean.columns

Index(['reportDate', 'symbol', 'reportedCurrency', 'cik', 'filingDate',
       'acceptedDate', 'calendarYear', 'period', 'cashAndCashEquivalents',
       'shortTermInvestments', 'cashAndShortTermInvestments', 'netReceivables',
       'inventory', 'otherCurrentAssets', 'totalCurrentAssets',
       'propertyPlantEquipmentNet', 'goodwill', 'intangibleAssets',
       'goodwillAndIntangibleAssets', 'longTermInvestments', 'taxAssets',
       'otherNonCurrentAssets', 'totalNonCurrentAssets', 'otherAssets',
       'totalAssets', 'accountPayables', 'shortTermDebt', 'taxPayables',
       'deferredRevenue', 'otherCurrentLiabilities', 'totalCurrentLiabilities',
       'longTermDebt', 'deferredRevenueNonCurrent',
       'deferredTaxLiabilitiesNonCurrent', 'otherNonCurrentLiabilities',
       'totalNonCurrentLiabilities', 'otherLiabilities',
       'capitalLeaseObligations', 'totalLiabilities', 'preferredStock',
       'commonStock', 'retainedEarnings',
       'accumulatedOtherComprehensiveIncomeLos

In [ ]:
balance_clean.shape

(47637, 62)

In [ ]:
(joined_data.infered_sec != joined_data.period).sum()

4927

In [ ]:
joined_data['infered_sec'] = np.nan
joined_data = joined_data.sort_values(['symbol','reportDate'])
# init joined_data on 4 if the form is a 10K statement
cond = joined_data.form.isin(['10-K', '10-K405', '10-KT'])
joined_data['infered_sec'] = joined_data['infered_sec'].mask(cond= cond, other=4)
# depth of 2
for i in range(2):
    # expand values vertically and horizontally
    for n in range(1,5):
        # only overwrite if its a nan
        nullcond = joined_data.infered_sec.isna()
        # move the 4 within group and make mask
        cond1 = joined_data.groupby('symbol').infered_sec.shift(n).eq(4)
        # set to value n if shifted mask & nan
        joined_data['infered_sec'].mask(cond=(cond1 & nullcond), other=n, inplace = True)
        # shift same for the other direction
        cond2 = joined_data.groupby('symbol').infered_sec.shift(-n).eq(4)
        if n != 4:
            joined_data['infered_sec'].mask(cond=(nullcond & cond2), other=(4-n), inplace = True)
        else:
            joined_data['infered_sec'].mask(cond=(nullcond & cond2), other=4, inplace = True)

reportDate symbol reportedCurrency      cik filingDate  \
33919 2021-09-30    OGN              USD  1821825 2021-11-12   
33920 2021-06-30    OGN              USD  1821825 2021-08-16   
33921 2021-03-31    OGN              USD  1821825 2021-06-21   
20    1992-12-31   ABMD              USD   815094 1992-12-31   
21    1992-09-30   ABMD              USD   815094 1992-09-30   
...          ...    ...              ...      ...        ...   
4032  2003-09-30    WRK              USD  1732845 2003-09-30   
4033  2003-06-30    WRK              USD  1732845 2003-06-30   
4034  2003-03-31    WRK              USD  1732845 2003-03-31   
4035  2002-12-31    WRK              USD  1732845 2002-12-31   
4067  1992-03-31   ZION              USD   109380 1992-03-31   

              acceptedDate  calendarYear  period  cashAndCashEquivalents  \
33919  2021-11-12 16:52:11          2021       3            1008000000.0   
33920  2021-08-16 08:20:13          2021       2             730000000.0   
33921  2021-06-21 08:16:50          2021       1             141000000.0   
20              1992-12-31          1993       3               3500000.0   
21              1992-09-30          1993       4               3200000.0   
...                    ...           ...     ...                     ...   
4032            2003-09-30          2003       4              14173000.0   
4033            2003-06-30          2003       3               8205000.0   
4034            2003-03-31          2003       2               5422000.0   
4035            2002-12-31          2003       1               6803000.0   
4067            1992-03-31          1992       2             468100000.0   

       shortTermInvestments  ...  reportDate_int  reportDate_month  \
33919                   0.0  ...        20210930                 9   
33920                   0.0  ...        20210630                 6   
33921                   0.0  ...        20210331                 3   
20                      0.0  ...        19921231                12   
21                      0.0  ...        19920930                 9   
...                     ...  ...             ...               ...   
4032                    0.0  ...        20030930                 9   
4033                    0.0  ...        20030630                 6   
4034                    0.0  ...        20030331                 3   
4035                    0.0  ...        20021231                12   
4067                    0.0  ...        19920331                 3   

       reportDate_year  reportDate_yday  filingDate_int  filingDate_month  \
33919             2021              273        20211112                11   
33920             2021              181        20210816                 8   
33921             2021               90        20210621                 6   
20                1992              365        19921231                12   
21                1992              274        19920930                 9   
...                ...              ...             ...               ...   
4032              2003              273        20030930                 9   
4033              2003              181        20030630                 6   
4034              2003               90        20030331                 3   
4035              2002              365        20021231                12   
4067              1992               91        19920331                 3   

       filingDate_year  filingDate_yday  form  infered_sec  
33919             2021              316  10-Q          NaN  
33920             2021              228  10-Q          NaN  
33921             2021              172  10-Q          NaN  
20                1992              365   NaN          NaN  
21                1992              274   NaN          NaN  
...                ...              ...   ...          ...  
4032              2003              273   NaN          NaN  
4033              2003              181   NaN          NaN  
4034     

In [ ]:
fred_daily = df.reset_index()
fred_daily.columns = real_names 

In [ ]:
# FRED SERIES
##########################################
# Daily indicators
series_names = ['DGS3MO', 
                'DGS1', 
                'DGS2', 
                'DGS3', 
                'DGS5', 
                'DGS10', 
                'DGS20', 
                'DGS30', 
                'DCOILWTICO', 
                'DCOILBRENTEU', 
                'VIXCLS',
                'GOLDAMGBD228NLBM']
real_names = ['Date',
              'T3m', 
              'T1y', 
              'T2y', 
              'T3y', 
              'T5y', 
              'T10y', 
              'T20y', 
              'T30y',
              'WTI', 
              'BRENT', 
              'VIX', 
              'GOLD']

start = datetime.datetime(1980, 1, 1)
end = datetime.date.today()

for i, name in enumerate(series_names):
    if i == 0:
        df = web.DataReader(name, data_source = 'fred', start=start, end = end)
    else:
        series = web.DataReader(name, data_source = 'fred', start=start, end = end)
        df = pd.concat([df, series], axis = 1)

fred_daily = df.reset_index()
fred_daily.columns = real_names 

# Exploration

In [ ]:
cflow = pd.read_csv('../data/raw/cflow_y.csv', index_col = 0)
income = pd.read_csv('../data/raw/income_y.csv', index_col = 0)
balance = pd.read_csv('../data/raw/balance_y.csv', index_col = 0)

## 10K Statements

special interest in period, symbol and calendarYear

In [ ]:
mycols = [
    'period', 
    'symbol', 
    'calendarYear'
    ]
balance[mycols].isna().sum()

period          1
symbol          0
calendarYear    0
dtype: int64

In [ ]:
balance[mycols].nunique()

period            5
symbol          505
calendarYear     38
dtype: int64

duplicate observations??

In [ ]:
# features to find duplicates
dup_features = [
    "symbol", 
    "calendarYear", 
    "totalAssets",
    "totalLiabilities", 
    "totalEquity",
    ]
mask_cols = dup_features + ["date", "fillingDate", "period"]
df_duplicates = check_duplicates(balance, dup_features).loc[:,mask_cols]


In [ ]:
df_duplicates

symbol  calendarYear     totalAssets  totalLiabilities    totalEquity  \
745      AMD          2013    3797000000.0      3382000000.0    415000000.0   
746      AMD          2013    3797000000.0      3382000000.0    415000000.0   
808      AMD          1997    3426000000.0      1382000000.0   2044000000.0   
810      AMD          1997    3426000000.0      1382000000.0   2044000000.0   
1011     AES          1996    3622000000.0      2688000000.0    934000000.0   
1014     AES          1996    3622000000.0      2688000000.0    934000000.0   
2247      MO          2021   48776000000.0     45781000000.0   2995000000.0   
2248      MO          2021   48776000000.0     45781000000.0   2995000000.0   
2581     AAL          2021   68437000000.0     75874000000.0  -7437000000.0   
2582     AAL          2021   68437000000.0     75874000000.0  -7437000000.0   
2630     AAL          2009   25754000000.0     28613000000.0  -2859000000.0   
2631     AAL          2009   25754000000.0     28613000000.0  -2859000000.0   
2635     AAL          2008   26950000000.0     26016000000.0    934000000.0   
2636     AAL          2008   26950000000.0     26016000000.0    934000000.0   
2638     AAL          2008   28766000000.0     26277000000.0   2489000000.0   
2639     AAL          2008   28766000000.0     26277000000.0   2489000000.0   
3263     AMP          2021  170726000000.0    165019000000.0   5707000000.0   
3265     AMP          2021  170726000000.0    165019000000.0   5707000000.0   
8185    BF-B          2020    6350000000.0      3947000000.0   2403000000.0   
8186    BF-B          2020    6350000000.0      3947000000.0   2403000000.0   
14255     DE          2021   80688000000.0     64955000000.0  15733000000.0   
14256     DE          2021   80688000000.0     64955000000.0  15733000000.0   
16838     EA          2007    5043000000.0      1165000000.0   3878000000.0   
16839     EA          2007    5043000000.0      1165000000.0   3878000000.0   
17665    ESS          2006    2459490000.0      1493131000.0    966359000.0   
17666    ESS          2006    2459490000.0      1493131000.0    966359000.0   
17667    ESS          2006    2328244000.0      1497768000.0    830476000.0   
17668    ESS          2006    2328244000.0      1497768000.0    830476000.0   
17669    ESS          2006    2296770000.0      1488470000.0    808300000.0   
17670    ESS          2006    2296770000.0      1488470000.0    808300000.0   
24304    IBM          2014  122646000000.0    105906000000.0  16740000000.0   
24305    IBM          2014  122646000000.0    105906000000.0  16740000000.0   
26723     KR          2022   48811000000.0     39582000000.0   9229000000.0   
26724     KR          2022   48811000000.0     39582000000.0   9229000000.0   
36166    PLD          2003    4976264000.0      2212085000.0   2764179000.0   
36167    PLD          2003    4976264000.0      2212085000.0   2764179000.0   
37027   QCOM          2004   10820000000.0      1156000000.0   9664000000.0   
37028   QCOM          2004   10820000000.0      1156000000.0   9664000000.0   
37032   QCOM          2003    8822436000.0      1223814000.0   7598622000.0   
37033   QCOM          2003    8822436000.0      1223814000.0   7598622000.0   
39081    STX          2012    9171000000.0      5706000000.0   3465000000.0   
39082    STX          2012    9171000000.0      5706000000.0   3465000000.0   
39784    SNA          2003    2138500000.0      1127600000.0   1010900000.0   
39785    SNA          2003    2138500000.0      1127600000.0   1010900000.0   
44628    VFC          2017    9173728000.0      5529389000.0   3644339000.0   
44629    VFC          2017    9173728000.0      5529389000.0   3644339000.0   
44630    VFC          2017    9246278000.0      4873221000.0   4373057000.0   
44631    VFC          2017    9246278000.0      4873221000.0   4373057000.0   
44846   VTRS          2020   31243100000.0     18695200000.0  12547900000.0   
44847   VTRS          2020   31243100000.0     186952000

In [ ]:
mask = (balance.period.isna()) | (balance.period == 'Q')
balance[mask]

date symbol reportedCurrency      cik fillingDate  \
4794  2010-02-04    ADM              USD     7084  2010-02-08   
9479  2001-09-30   CBRE              USD  1138118  2001-11-14   

             acceptedDate  calendarYear period  cashAndCashEquivalents  \
4794  2010-02-05 19:10:55          2010      Q            1317000000.0   
9479  2001-11-14 00:00:00          2001   <NA>              24211000.0   

      shortTermInvestments  ...  totalStockholdersEquity  \
4794           317000000.0  ...            14744000000.0   
9479                   0.0  ...              247463000.0   

      totalLiabilitiesAndStockholdersEquity  minorityInterest    totalEquity  \
4794                          23878000000.0      5987000000.0  20731000000.0   
9479                           1219638000.0               0.0    247463000.0   

      totalLiabilitiesAndTotalEquity  totalInvestments    totalDebt  \
4794                   29865000000.0      3661000000.0  732000000.0   
9479                    1219638000.0               0.0  639664000.0   

          netDebt                                               link  \
4794 -585000000.0  https://www.sec.gov/Archives/edgar/data/7084/0...   
9479  615453000.0  https://www.sec.gov/Archives/edgar/data/113811...   

                                              finalLink  
4794  https://www.sec.gov/Archives/edgar/data/7084/0...  
9479  https://www.sec.gov/Archives/edgar/data/113811...  

[2 rows x 54 columns]

In [ ]:
print(balance.loc[mask, mask_cols].to_markdown())

|      | symbol   |   calendarYear |   totalAssets |   totalLiabilities |   totalEquity | date       | fillingDate   | period   |
|-----:|:---------|---------------:|--------------:|-------------------:|--------------:|:-----------|:--------------|:---------|
| 4794 | ADM      |           2010 |   2.9865e+10  |        9.134e+09   |   2.0731e+10  | 2010-02-04 | 2010-02-08    | Q        |
| 9479 | CBRE     |           2001 |   1.21964e+09 |        9.72175e+08 |   2.47463e+08 | 2001-09-30 | 2001-11-14    | <NA>     |


Are there any observations where, for a given symbol and year, does not have 4 periods?? First and last year excluded

In [ ]:
# get min and max from each group and filter them out
min_year = balance.groupby('symbol')['calendarYear'].transform('min')
max_year = balance.groupby('symbol')['calendarYear'].transform('max')
filter = (balance.calendarYear == min_year) | (balance.calendarYear == max_year)
# get nunique periods for each calendar year excluding min and max years
aux = balance[~filter].groupby(['symbol', 'calendarYear']).agg({'period': ['nunique']}).reset_index()
aux.columns = ('symbol', 'calendarYear', 'period_nunique')
# show in balance the values which dont have 4 unique periods per year
df_incomplete = aux[aux.period_nunique!=4]
keys = tuple(zip(df_incomplete.symbol, df_incomplete.calendarYear))
# creating the mask
mask = False
for symbol, year in keys:
    mask = mask | ((balance.symbol==symbol) & (balance.calendarYear==year))
# dataframe with all the observations
df_incomplete = balance.loc[mask, :].sort_values(by=['symbol', 'calendarYear', 'period'])


In [ ]:
balance.loc[[902,903,904], mask_cols]

symbol  calendarYear   totalAssets  totalLiabilities  totalEquity  \
902    AAP          2004  2063000000.0      1374000000.0  689000000.0   
903    AAP          2004  1983000000.0      1352000000.0  631000000.0   
904    AAP          2003  2011000000.0      1414000000.0  597000000.0   

           date fillingDate period  
902  2004-04-24  2004-05-21     Q1  
903  2004-01-03  2004-03-12     Q4  
904  2003-10-04  2003-11-06     Q3

In [ ]:
numerical = balance.select_dtypes(include=['int', 'float']).columns
explore_numerical(balance, numerical)

nulls  nulls_pct           min  \
cik                                          0        0.0  1.800000e+03   
calendarYear                                 0        0.0  1.985000e+03   
cashAndCashEquivalents                       0        0.0 -3.130000e+07   
shortTermInvestments                         0        0.0 -4.284000e+09   
cashAndShortTermInvestments                  0        0.0 -2.874000e+09   
netReceivables                               0        0.0 -1.418230e+08   
inventory                                    0        0.0 -1.783500e+07   
otherCurrentAssets                           0        0.0 -2.264398e+17   
totalCurrentAssets                           0        0.0 -2.264398e+17   
propertyPlantEquipmentNet                    0        0.0 -1.457000e+07   
goodwill                                     0        0.0 -1.091000e+09   
intangibleAssets                             0        0.0 -2.233000e+09   
goodwillAndIntangibleAssets                  0        0.0 -3.990000e+08   
longTermInvestments                          0        0.0 -1.311697e+09   
taxAssets                                    0        0.0  0.000000e+00   
otherNonCurrentAssets                        0        0.0 -1.229959e+12   
totalNonCurrentAssets                        0        0.0 -1.218035e+12   
otherAssets                                  0        0.0  0.000000e+00   
totalAssets                                  0        0.0 -1.808275e+09   
accountPayables                              0        0.0 -3.473100e+07   
shortTermDebt                                0        0.0 -6.786000e+09   
taxPayables                                  0        0.0 -2.231000e+07   
deferredRevenue                              0        0.0 -2.360000e+09   
otherCurrentLiabilities                      0        0.0 -1.177565e+12   
totalCurrentLiabilities                      0        0.0  0.000000e+00   
longTermDebt                                 0        0.0 -5.761100e+07   
deferredRevenueNonCurrent                    0        0.0  0.000000e+00   
deferredTaxLiabilitiesNonCurrent             0        0.0 -8.992000e+09   
otherNonCurrentLiabilities                   0        0.0 -3.908000e+11   
totalNonCurrentLiabilities                   0        0.0 -1.807700e+10   
otherLiabilities                             0        0.0  0.000000e+00   
capitalLeaseObligations                      0        0.0  0.000000e+00   
totalLiabilities                             0        0.0 -5.930000e+09   
preferredStock                               0        0.0  0.000000e+00   
commonStock                                  0        0.0 -4.340100e+10   
retainedEarnings                             0        0.0 -3.407600e+10   
accumulatedOtherComprehensiveIncomeLoss      0        0.0 -1.026420e+11   
othertotalStockholdersEquity                 0        0.0 -3.634644e+11   
totalStockholdersEquity                      0        0.0 -2.556000e+10   
totalLiabilitiesAndStockholdersEquity        0        0.0 -2.598300e+10   
minorityInterest                             0        0.0 -1.907432e+12   
totalEquity                                  0        0.0 -1.720214e+12   
totalLiabilitiesAndTotalEquity               0        0.0 -1.808275e+09   
totalInvestments                             0        0.0 -4.284000e+09   
totalDebt                                    0        0.0 -4.392000e+06   
netDebt                                      0        0.0 -2.849530e+11   

                                                  max          mean  \
cik                                      1.841666e+06  6.807326e+05   
calendarYear                             2.022000e+03  2.008682e+03   
cashAndCashEquivalents                   7.598690e+11  3.144459e+09   
shortTermInvestments                     4.830690e+11  3.485959e+09   
cashAndShortTermInvestments              1.026474e+12  6.630419e+09   
netReceivables                           4.538370e+11  3.238153e+09   
inventory         

### Period Inconsistencies
1. Misnomer periods
<br>

|      | symbol   |   calendarYear |   totalAssets |   totalLiabilities |   totalEquity | date       | fillingDate   | period   |
|-----:|:---------|---------------:|--------------:|-------------------:|--------------:|:-----------|:--------------|:---------|
| 4794 | ADM      |           2010 |   2.9865e+10  |        9.134e+09   |   2.0731e+10  | 2010-02-04 | 2010-02-08    | Q        |
| 9479 | CBRE     |           2001 |   1.21964e+09 |        9.72175e+08 |   2.47463e+08 | 2001-09-30 | 2001-11-14    | <NA>     |

2. Periods carry to the next year as in index=903
<br>

|     | symbol   |   calendarYear |   totalAssets |   totalLiabilities |   totalEquity | date                | fillingDate         | period   |
|----:|:---------|---------------:|--------------:|-------------------:|--------------:|:--------------------|:--------------------|:---------|
| 902 | AAP      |           2004 |     2.063e+09 |          1.374e+09 |      6.89e+08 | 2004-04-24 00:00:00 | 2004-05-21 00:00:00 | Q1       |
| 903 | AAP      |           2004 |     1.983e+09 |          1.352e+09 |      6.31e+08 | 2004-01-03 00:00:00 | 2004-03-12 00:00:00 | Q4       |
| 904 | AAP      |           2003 |     2.011e+09 |          1.414e+09 |      5.97e+08 | 2003-10-04 00:00:00 | 2003-11-06 00:00:00 | Q3       |

2. Duplicates rows with different periods
<br>

|      | symbol   |   calendarYear |   totalAssets |   totalLiabilities |   totalEquity | date                | fillingDate         | period   |
|-----:|:---------|---------------:|--------------:|-------------------:|--------------:|:--------------------|:--------------------|:---------|
| 2581 | AAL      |           2021 |    6.8437e+10 |         7.5874e+10 |    -7.437e+09 | 2021-09-30 00:00:00 | 2021-10-21 00:00:00 | Q3       |
| 2582 | AAL      |           2021 |    6.8437e+10 |         7.5874e+10 |    -7.437e+09 | 2021-09-03 00:00:00 | 2021-10-21 00:00:00 | Q4       |



